### PR TITLE
Improving Code quality of JsonMessage.java

### PR DESCRIPTION
**Pull request by cb0s**
- Removing duplicate code parts from Json-Message
- Update .gitignore with venv/ for python

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -301,3 +301,6 @@ $RECYCLE.BIN/
 
 # Project Files
 modules/telestion-mavlink/src/main/messages/org/telestion/mavlink/
+
+# Python things
```
Comment (jvpichowski) [inline .gitignore:305]: This change should better go into its own PR.

File: modules/telestion-api/src/main/java/org/telestion/api/message/JsonMessage.java [hunk 2]
```diff
@@ -1,24 +1,26 @@
...
+import org.slf4j.LoggerFactory;
 
 /**
  * The base class for all messages which are automatically encoded with the JsonMessageCodec.<br>
  * All subclasses have to be valid json classes. This means that they could be encoded by
  * {@link io.vertx.core.spi.json.JsonCodec} which is backed by {@link io.vertx.core.json.jackson.JacksonCodec}.
  *
  * @author Jan von Pichowski, Cedric Boes
- * @version 1.1
+ * @version 1.2
```
Comment (jvpichowski) [inline modules/telestion-api/src/main/java/org/telestion/api/message/JsonMessage.java:18]: It is okay for now but we have to rework the versioning scheme.

File: modules/telestion-api/src/main/java/org/telestion/api/message/JsonMessage.java [hunk 3]
```diff
@@ -117,4 +99,28 @@ default String className() {
...
+	/**
+	 * Reflection part of {@link #on(Class, Message, Handler)} and {@link #on(Class, Object, Handler)}.
+	 *
+	 * @param clazz			Class to cast to
+	 * @param jsonObject	Object to be casted
+	 * @param handler		Handler for the message
+	 * @param <T>			Type of the message
+	 * @return if parsing successful
+	 */
+	private static <T extends JsonMessage> boolean doReflection(Class<T> clazz, JsonObject jsonObject,
```
Comment (jvpichowski) [inline modules/telestion-api/src/main/java/org/telestion/api/message/JsonMessage.java:112]: The name is not very precise. While it is generally a good idea to pull out duplicated code you don't necessarily need to introduce a new method. Here it would be better to chain the method call in the second variant with a call like `return on(msg.body(),...)`. You don't need this new method.

File: modules/telestion-api/src/main/java/org/telestion/api/message/JsonMessage.java [hunk 4]
```diff
@@ -117,4 +99,28 @@ default String className() {
...
+																Handler<T> handler) {
+		try {
+			var msgClazz = Class.forName(jsonObject.getString("className"));
+			if (!clazz.isAssignableFrom(msgClazz)) {
+				return false;
+			}
+			handler.handle((clazz.cast(jsonObject.mapTo(msgClazz))));
+			return true;
+		} catch (ClassNotFoundException e) {
+			logger.error("Error while converting Json into JsonMessage!", e);
```
Comment (jvpichowski) [inline modules/telestion-api/src/main/java/org/telestion/api/message/JsonMessage.java:122]: Please try to avoid the exclamation mark at the end of a log message and be more precise with the message. You need to cover the reason to provide more information. Something like "the requested message class is not available on the classpath".

The error level is way too high here. I think info or warning would be more appropriate. Think about it this way: We expected a certain type of class that is clearly on the classpath but we got a message which is not on the path. So we don't really care about this message. But it could be nice to know, that the system is receiving unknown messages. Maybe trash?

**Review by jvpichowski — CHANGES_REQUESTED** [review]
Good catch! This could be simplified. Please consider my notes to improve it further.

File: modules/telestion-api/src/main/java/org/telestion/api/message/JsonMessage.java (hunk 4 above)
Comment (cb0s) [reply modules/telestion-api/src/main/java/org/telestion/api/message/JsonMessage.java:122]: Wasn't my code. Thought you did this on purpose :)

Comment (jvpichowski) [reply modules/telestion-api/src/main/java/org/telestion/api/message/JsonMessage.java:122]: Then it is better to rework this.

File: .gitignore (hunk 1 above)
Comment (cb0s) [reply .gitignore:305]: Will do in the future :D

File: modules/telestion-api/src/main/java/org/telestion/api/message/JsonMessage.java [hunk 5]
```diff
@@ -117,4 +99,30 @@ default String className() {
...
+	 * Reflection part of {@link #on(Class, Message, Handler)} and {@link #on(Class, Object, Handler)} which constructs
+	 * the original message.
+	 *
+	 * @param clazz			Class to cast to
+	 * @param jsonObject	Object to be casted
+	 * @param handler		Handler for the message
+	 * @param <T>			Type of the message
+	 * @return if parsing successful
+	 */
+	private static <T extends JsonMessage> boolean reconstructMsg(Class<T> clazz, JsonObject jsonObject,
```
Comment (jvpichowski) [inline modules/telestion-api/src/main/java/org/telestion/api/message/JsonMessage.java:113]: Better method name, but you still don't need this method.

File: modules/telestion-api/src/main/java/org/telestion/api/message/JsonMessage.java [hunk 6]
```diff
@@ -117,4 +99,30 @@ default String className() {
...
+																  Handler<T> handler) {
+		try {
+			var msgClazz = Class.forName(jsonObject.getString("className"));
+			if (!clazz.isAssignableFrom(msgClazz)) {
+				return false;
+			}
+			handler.handle((clazz.cast(jsonObject.mapTo(msgClazz))));
+			return true;
+		} catch (ClassNotFoundException e) {
+			logger.error("Error while converting JSON into JsonMessage. The requested message is not available on the" +
```
Comment (jvpichowski) [inline modules/telestion-api/src/main/java/org/telestion/api/message/JsonMessage.java]: Again please reduce the log level.

**Review by jvpichowski — CHANGES_REQUESTED** [review]
Better but still some improvements

File: modules/telestion-api/src/main/java/org/telestion/api/message/JsonMessage.java [hunk 7]
```diff
@@ -39,7 +41,8 @@
 				handler.handle((clazz.cast(jsonObject.mapTo(msgClazz))));
 				return true;
 			} catch (ClassNotFoundException e) {
-				e.printStackTrace();
+				logger.error("Error while converting JSON into JsonMessage. The requested message is not available on the "
```
Comment (jvpichowski) [inline modules/telestion-api/src/main/java/org/telestion/api/message/JsonMessage.java]: Please reduce this to warn. Then it is fine :)

**Review by jvpichowski — CHANGES_REQUESTED** [review]
Just one more thing :D

File: modules/telestion-api/src/main/java/org/telestion/api/message/JsonMessage.java (hunk 7 above)
Comment (cb0s) [reply modules/telestion-api/src/main/java/org/telestion/api/message/JsonMessage.java]: But isn't this a critical error?

Comment (jvpichowski) [reply modules/telestion-api/src/main/java/org/telestion/api/message/JsonMessage.java]: No, it is not. See my notes above. It is just a hint that somebody is dropping some unknown content at you. No need to worry too much.

**Review by jvpichowski — APPROVED** [review]
It is fine. Go for it!
